test/challtestsrv/README.md

```diff
@@ -0,0 +1,117 @@
+  -http01 string
+       Comma separated bind addresses/ports for HTTP-01 challenges. Set empty to disable. (default ":5002")
+  -management string
+       Bind address/port for management HTTP interface (default ":8056")
+```
+
+To disable a challenge type, set the bind address to `""`. E.g.:
+
+* To run HTTP-01 only: `challtestsrv -dns01 ""`
+* To run DNS-01 only: `challtestsrv -http01 ""`
+
+### Management Interface
+
+_Note: These examples assume the default management interface of `:8056`_
+
+Adding an HTTP-01 challenge response for the token `"aaaa"` with the content
+`"bbbb"`:
+
+    curl -X POST -d '{"token":"aaaa", "content":"bbbb"}' localhost:8056/add-http01
+
+Deleting an HTTP-01 challenge response for the token `"aaaa"`:
+
+    curl -X POST -d '{"token":"aaaa"}' localhost:8056/del-http01
+
+Adding a DNS-01 TXT challenge for the host `"_acme-challenge.example.com."`
+with the value `"bbbb"`:
+
+    curl -X POST -d '{"host":"_acme-challenge.example.com.", "value":"bbbb"}' localhost:8056/set-txt
+
+Deleting a DNS-01 TXT challenge for the host `"_acme-challenge.example.com."`:
+
+    curl -X POST -d '{"host":"_acme-challenge.example.com."}' localhost:8056/clear-txt
+
+## The `test/challtestsrv` package
+
+The `test/challtestsrv` package can be used as a library by another program to
+avoid needing to manage an external `challtestsrv` binary or use the HTTP based
+management interface. This is used by the Boulder `load-generator` command to
+manage its own in-process HTTP-01 challenge server.
+
+### Usage
+
+Create a challenge server responding to HTTP-01 challenges on ":8888" and
+DNS-01 challenges on ":9999" and "10.0.0.1:9998":
+
+```
+  import "github.com/letsencrypt/boulder/test/challtestsrv"
+
+  challSrv, err := challtestsrv.New(challsrv.Config{
+    HTTPOneAddr: []string{":8888"},
+    DNSOneAddr: []string{":9999", "10.0.0.1:9998"},
+  })
+  if err != nil {
+    panic(err)
+  }
+```
+
+Run the Challenge server:
+```
+  // Create a waitgroup that can be used to block until the challenge server has
+  // cleanly shut down
+  challSrvWg := new(sync.WaitGroup)
+  challSrvWg.Add(1)
+  // Start the Challenge server in its own Go routine
+  go challSrv.Run(challSrvWg)
+```
+
+Add an HTTP-01 response for the token `"aaa"` and the value `"bbb"`, defer
+cleaning it up again:
+```
+  challSrv.AddHTTPOneChallenge("aaa", "bbb")
+  defer challSrv.DeleteHTTPOneChallenge("aaa")
+```
+
+Add a DNS-01 TXT response for the host `"_acme-challenge.example.com."` and the
+value `"bbb"`, defer cleaning it up again:
+```
+  challSrv.AddDNSOneChallenge("_acme-challenge.example.com.", "bbb")
+  defer challSrv.DeleteHTTPOneChallenge("_acme-challenge.example.com.")
+```
+
+Stop the Challenge server:
+```
+  // Send a Shutdown request to the challenge server
+  challSrv.Shutdown()
+  // Wait on the waitgroup we gave the challenge server when we called Run().
```

Comment:
I think `challSrv` should create its own waitGroup and block on it when Shutdown is called. That matches the behavior of `net/http.Server` more closely.